File: VocabLarry/PHP/auth/mailer.php
<?php
function get_mail_config(): ?array {
    $path = __DIR__ . '/config.local.php';
    if (!file_exists($path)) return null;
    return require $path;
}

function smtp_read($socket): string {
    $data = '';
    while (($line = fgets($socket, 515)) !== false) {
        $data .= $line;
        if (isset($line[3]) && $line[3] === ' ') break;
    }
    return $data;
}

function smtp_send_mail(string $to, string $subject, string $bodyText): bool {
    $config = get_mail_config();
    if (!$config) {
        error_log('Mailer: auth/config.local.php is missing, cannot send email.');
        return false;
    }

    $socket = @stream_socket_client(
        'tcp://' . $config['smtp_host'] . ':' . $config['smtp_port'],
        $errno, $errstr, 15
    );
    if (!$socket) {
        error_log("Mailer: connection failed: $errstr");
        return false;
    }

    smtp_read($socket);
    fwrite($socket, "EHLO localhost\r\n");
    smtp_read($socket);

    fwrite($socket, "STARTTLS\r\n");
    smtp_read($socket);
    if (!stream_socket_enable_crypto($socket, true, STREAM_CRYPTO_METHOD_TLS_CLIENT)) {
        error_log('Mailer: TLS negotiation failed');
        fclose($socket);
        return false;
    }

    fwrite($socket, "EHLO localhost\r\n");
    smtp_read($socket);

    fwrite($socket, "AUTH LOGIN\r\n");
    smtp_read($socket);
    fwrite($socket, base64_encode($config['smtp_user']) . "\r\n");
    smtp_read($socket);
    fwrite($socket, base64_encode($config['smtp_pass']) . "\r\n");
    $authResp = smtp_read($socket);
    if (strpos($authResp, '235') !== 0) {
        error_log("Mailer: auth failed: $authResp");
        fclose($socket);
        return false;
    }

    fwrite($socket, "MAIL FROM:<{$config['from_email']}>\r\n");
    smtp_read($socket);
    fwrite($socket, "RCPT TO:<{$to}>\r\n");
    smtp_read($socket);
    fwrite($socket, "DATA\r\n");
    smtp_read($socket);

    $headers  = "From: {$config['from_name']} <{$config['from_email']}>\r\n";
    $headers .= "To: <{$to}>\r\n";
    $headers .= "Subject: {$subject}\r\n";
    $headers .= "MIME-Version: 1.0\r\n";
    $headers .= "Content-Type: text/plain; charset=UTF-8\r\n";

    $escapedBody = preg_replace('/^\./m', '..', $bodyText);
    fwrite($socket, $headers . "\r\n" . $escapedBody . "\r\n.\r\n");
    smtp_read($socket);

    fwrite($socket, "QUIT\r\n");
    smtp_read($socket);
    fclose($socket);

    return true;
}

File: IELTS-Vocab/Python/Flask/auth/mailer.py
import importlib.util
import logging
import os
import smtplib
from email.mime.text import MIMEText


def get_mail_config() -> dict | None:
    path = os.path.join(os.path.dirname(__file__), 'config.local.py')
    if not os.path.exists(path):
        return None
    spec = importlib.util.spec_from_file_location('config_local', path)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return getattr(mod, 'CONFIG', None)


def smtp_send_mail(to: str, subject: str, body_text: str) -> bool:
    config = get_mail_config()
    if not config:
        logging.error('Mailer: auth/config.local.py is missing, cannot send email.')
        return False

    msg = MIMEText(body_text, 'plain', 'utf-8')
    msg['From'] = f"{config['from_name']} <{config['from_email']}>"
    msg['To'] = to
    msg['Subject'] = subject

    try:
        with smtplib.SMTP(config['smtp_host'], config['smtp_port'], timeout=15) as smtp:
            smtp.ehlo()
            smtp.starttls()
            smtp.ehlo()
            smtp.login(config['smtp_user'], config['smtp_pass'])
            smtp.sendmail(config['from_email'], [to], msg.as_string())
        return True
    except Exception as exc:
        logging.error('Mailer: %s', exc)
        return False
